### PR TITLE
Add basic wildcard/keyword fallback for upcoming ECS releases

### PR DESCRIPTION
**Pull request by andrewstucki**
## What does this PR do?

This sets the groundwork for https://github.com/elastic/beats/issues/21674 by adding a simple fallback mechanism for wildcard fields in unsupported Elasticsearch versions and non-OSS Beats distros. If you're publishing to ES < 7.9.0 you get `keyword` fields instead of wildcard, if you're coming from OSS Beats you get the same.

This will allow us to move forward with the migration to ECS 1.7.0 with upcoming wildcard support..

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: libbeat/version/helper.go [hunk 1]
```diff
@@ -32,6 +32,9 @@ var (
 	buildTime = "unknown"
 	commit    = "unknown"
 	qualifier = ""
+
+	// OSS provides if this is the OSS distro
+	OSS = true
```
Comment (andrewstucki) [inline libbeat/version/helper.go:37]: the global with the `init` override from the xpack root command is kind of sick, but I'm not sure there's a simpler/cleaner way of doing this

**Comment by webmat** [comment]
This is looking good for now. The next version of ECS will only introduce the `wildcard` basic type.

I'm not familiar enough about the codebase, but it's important that this fallback happens for any field using basic types, not just ECS fields. I don't have insight into what custom fields are in Beats or will be added, but if any of them is basic or changes to a basic type, they will need to fall back to an OSS type as well, where possible.

As a corollary to this previous point, perhaps you should consider falling back `constant_keyword` and `version` to `keyword` as well?

* For reasons I won't go into, ECS will likely not introduce `constant_keyword` to the Beats field definitions. But if each Beat decided to migrate fields that remain constant in the current indexing strategy like `agent.type` to `constant_keyword`, a fallback would be necessary.
* At this time `version` looks like a conflict with `keyword`, so it's unclear if any pre-existing fields will migrate to it during 7.x. But this data type is fair game for any net new field in ECS or custom fields. So being ready to perform this fallback sounds like a good idea.

So to recap, as far as the next version of ECS is concerned, only `wildcard` => `keyword` is necessary, and this seems to do it 👍  But make sure the fallback happens on all fields, not just ECS ones; and make sure to confirm whether any more fallbacks are necessary based on Beats field definitions that aren't driven by ECS.

**Comment by andrewstucki** [comment]
@webmat my understanding of this code is that the `Processor` gets called on the final merged beats fields that are going into the Elasticsearch index template, meaning it should include both any fields specified at the module level, as well as the global ECS fields yml file that we include in `libbeat/_meta/fields.ecs.yml`. Just to keep the scope of the change smaller for now, I'm thinking it probably makes sense to add the `constant_keyword` and `version` fallbacks as they crop up, but if you guys really want me to, I can add it here too.

**Comment by webmat** [comment]
Not required per ECS, so feel free to move forward as is.

But as I said, I don't have insight into all of the other Beats fields. Trying each OSS Beats' `...beat setup` command on an OSS cluster might be a quick way find out if there's other basic types that have been added.

**Review by marc-gr — APPROVED** [review]

File: x-pack/libbeat/cmd/inject.go [hunk 2]
```diff
@@ -21,6 +22,10 @@ import (
 	_ "github.com/elastic/beats/v7/x-pack/libbeat/autodiscover/providers/aws/elb"
 )
 
+func init() {
+	version.ElasticLicensed = true
```
Comment (exekias) [inline x-pack/libbeat/cmd/inject.go:26]: I would expect this to be retrieved from the Elasticsearch cluster when we check for the version? The same response that is providing the version gives us the Elasticsearch flavour, see `build_flavor` here, which can be `default` or `oss`:

```
{
  "name" : "01e1731f6d10",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "h3NzNl8MSvyhSyNoMGreiw",
  "version" : {
    "number" : "7.10.0-SNAPSHOT",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "6be4acd5094954851a956dd93d8c4043979011d9",
    "build_date" : "2020-11-08T20:00:29.019941Z",
    "build_snapshot" : true,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

The code that is currently retrieving version info is here:
https://github.com/elastic/beats/blob/d97fe08d6d5bb27e4c5e52d04a45448fd1fe8827/libbeat/esleg/eslegclient/connection.go#L402, check for the `Ping` implementation

Comment (exekias) [reply x-pack/libbeat/cmd/inject.go:26]: Ah sorry, I missed the point that this is not needed

Comment (exekias) [inline x-pack/libbeat/cmd/inject.go:26]: This is setting a global, that we are trying to avoid cc @urso

Perhaps we should come up with a way to passing this to the beat initialization?

Comment (urso) [reply x-pack/libbeat/cmd/inject.go:26]: Yeah, +1 on not introducing globals. Beats query the ES version and Basic queries the license on startup.

We pass the agent name and version via `beat.Info`. Maybe this would be the right place to put this information.

Comment (andrewstucki) [reply x-pack/libbeat/cmd/inject.go:26]: I'm not a fan of globals either, and made the change, but due to the way that the root cobra commands are initialized in each beat and the way the info propagates from the command to the processor, the changes are quite a bit more invasive. So, I'm going to keep this conversation unresolved for some context in case we want to revert back to the global.

File: libbeat/beat/info.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ type Info struct {
 	Beat        string    // The actual beat's name
 	IndexPrefix string    // The beat's index prefix in Elasticsearch.
 	Version     string    // The beat version. Defaults to the libbeat version when an implementation does not set a version
+	XPack       bool      // Whether the beat is licensed under xpack
```
Comment (exekias) [inline libbeat/beat/info.go:27]: This should be Elastic license, not xpack, how about `ElasticLicense`?

**Review by exekias — COMMENTED** [review]
We are almost there, I left a minor comment on naming, apart from that this LGTM! Thank you for working on it

File: libbeat/beat/info.go (hunk 3 above)
Comment (andrewstucki) [reply libbeat/beat/info.go:27]: Sure, I'll change the name back

**Review by exekias — APPROVED** [review]

**Comment by simitt** [comment]
cc @elastic/apm-server when pulling this change in with a beats update we have to ensure to set the `setting.ElasticLicense` accordingly.
